connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java

```diff
@@ -240,6 +261,21 @@ public Runnable withClassLoader(ClassLoader classLoader, Runnable operation) {
         };
     }
 
+    public static LoaderSwap swapLoader(ClassLoader loader) {
```

Comment:
I made this static so that I can directly invoke it from the runtime classes while swapping loaders for converters, transformations. Making this non-static is an option but we should then move this logic to the runtime (or pass the plugins instance to the WorkerTasks which seems overkill). I will lead to a small code duplication but will aid in mocking and testing there.

Reply:
So far the Worker has maintained the Plugins instance, and the WorkerTask has maintained the ClassLoader instance, allowing `plugins.swapLoader(task.loader())` to be called when necessary.

> pass the plugins instance to the WorkerTasks which seems overkill

I think this is the best way forward.

Reply:
I realize now that there's a lot of classloaders flying around inside a WorkerTask, and loader() really doesn't cover them all.

Your calls to swapLoader in convertAndTransformRecord are actually new, and fixes an existing TCCL bug.
Fix it fully would require ClassLoader fields for each converter/predicate/transform/etc.

We can either leave it unaddressed, or throw in some lazy plugin.swapLoader(plugin.connectorLoader(pluginName, version)) calls...

What a mess. This is exactly why I wanted to get https://issues.apache.org/jira/browse/KAFKA-14670 done :(

Reply:
Okay I thought about this some more.

The fact that this is static can be fixed later. But the fact that we're using plugin.getClass().getClassLoader() to get the ClassLoader is incorrect, and I think either needs a Plugins instance to call connectorLoader, or needs new ClassLoader fields anyway, so it may not be that much work to add the Plugins field as well.

The only thing we need to avoid is accidentally swapping to the AppClassLoader (parent of DelegatingClassLoader) instead of the DelegatingClassLoader itself, because that actually would make some classes less visible to plugins.